docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,552 @@
+Thus, in the above example, the call evaluates to 22.
+In the case of operators, the implementation is opaque to Relay,
+so the result is left up to the registered TVM implementation.
+
+.. *Note: type parameters are not yet supported in the text format.* 
+
+A type-polymorphic function can also include type arguments at a call
+site. The type arguments are substituted for type parameters when
+type checking. If a function is type-polymorphic and type arguments are not
+given, type inference will attempt to infer type arguments if possible.
+The following code gives examples of explicit and inferred type arguments:
+
+.. code-block:: python
+
+    // %f : fn<a : Type, b : Type, c : Type>(a, b) -> c
+    let %x1 = f<Tensor[(), bool], Tensor[(), bool], Tensor[(), bool)]>(True, False);
+    // %x1 is of type Tensor[(), bool]
+    let %x2 : () = f(%x1, %x1)
+    // the type arguments in the second call are inferred to be <Tensor[(), bool], Tensor[(), bool], ()>
+
+Note that all type relations in the function type must hold at each
+call site. Specifically, this means that the relation will be checked
+against the specific types of the arguments at a given call site. This 
+is also a form of polymorphism, since there may be multiple valid
+assignments of argument types and a return type so long as the relation
+is satisfied.
+
+For example, if we have a function :code:`%f` that takes tensor arguments
+and has the :code:`Broadcast` relation, then there are many different
+shapes that the arguments in the below call could have that would satisfy
+the type annotation:
+
+.. code-block:: python
+
+   let %x : Tensor[(100, 100, 100), float32] = %f(%a, %b, %c);
+   %x
+
+See :py:class:`~tvm.relay.expr.Call` for its definition and documentation.
+
+Module and Global Functions
+===========================
+
+Relay keeps a global data structure known as a "module" (often called an "environment" in other
+functional programming languages) to keep track of the definitions of global functions.
+In particular, the module keeps a globally accessible mapping of global variables to the
+function expressions they denote. The utility of the module is that it allows global functions
+to recursively refer to themselves or any other global function (e.g., as in mutual recursion).
+
+Note Relay's module is analogous to data structures for keeping track of subgraphs in computation
+graph-based IRs.
+
+Global functions in Relay behave identically to the function expressions defined in `Functions`_,
+but have syntactic sugar in the text format to enter their definitions into the module. Namely,
+a global function definition includes a global identifier and is allowed to recursively refer to
+that identifier in the body, as in the following example:
+
+.. code-block:: python
+
+   def @ackermann(%m : Tensor[(), int32], %n : Tensor[(), int32]) -> Tensor[(), int32] {
+       if (%m == 0) {
+           %n + 1
+       } else if (%m > 0 && n == 0) {
```

Comment:
`else if` (and `&&`?) is not supported right now, but let's keep it and add an issue

Reply:
I think they should be supported if not presently, yeah